### PR TITLE
Certified assets.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

### CODE REVIEW

File: wallet/src/lib.rs [hunk 1]
```diff
@@ -103,6 +109,108 @@ fn set_name(name: String) {
...
+    let parts: Vec<&str> = req.url.split('?').collect();
+    let asset = parts[0].trim_start_matches('/');
+    let asset = if asset.is_empty() {
+        "index.html"
+    } else {
+        asset
+    };
+    let assets = storage::get::<Assets>();
+    let certificate_header = make_asset_certificate_header(&assets.hashes, asset);
+    eprintln!("hi {}", asset);
```
Comment (hansl) [inline wallet/src/lib.rs:164]: Might want to remove the logging :D

Comment (jplevyak) [reply wallet/src/lib.rs]: done

File: wallet/src/lib.rs [hunk 2]
```diff
@@ -218,7 +324,7 @@ mod wallet {
             args.canister.clone(),
             "wallet_receive",
             (),
-            args.amount as i64,
+            args.amount as u64,
```
Comment (p-shahi) [inline wallet/src/lib.rs:327]:
```suggestion
            args.amount,
```
they're already u64 we can just drop the as i64 here

**Comment by p-shahi** [comment]
thanks for this! I haven't looked at the certified assets canister so you'll want Hans or @roman-kashitsyn's review here

File: wallet/src/lib.rs (hunk 2 above)
Comment (jplevyak) [reply wallet/src/lib.rs]: done

**Comment by jplevyak** [comment]
I am getting: validation.ts:58 Witness != Tree passed in ic-certification
fB @ validation.ts:58
http_request.ts:313 BODY DOES NOT PASS VERIFICATION
UB @ http_request.ts:313
2validation.ts:58 Witness != Tree passed in ic-certification
fB @ validation.ts:58
http_request.ts:313 BODY DOES NOT PASS VERIFICATION
UB @ http_request.ts:313

@roman-kashitsyn seems like I have a bug.

File: dfx.json [hunk 3]
```diff
@@ -23,7 +23,15 @@
...
-      "providers": ["https://gw.dfinity.network"],
+      "providers": ["https://ic0.app"],
+      "type": "persistent"
+    },
+    "consensus1": {
+      "providers": ["https://consensus1.dfinity.network"],
+      "type": "persistent"
+    },
+    "consensus2": {
+      "providers": ["https://consensus2.dfinity.network"],
```
Comment (p-shahi) [inline dfx.json:34]:
```suggestion
```

File: wallet/Cargo.toml [hunk 4]
```diff
@@ -13,11 +13,17 @@ crate-type = ["cdylib"]
 path = "src/lib.rs"
 
 [dependencies]
+base64 = "0.13.0"
 ic-types = "0.1.2"
-ic-cdk = "0.2.3"
-ic-cdk-macros = "0.2.3"
+ic-cdk = { git = "https://github.com/dfinity/cdk-rs", branch = "main"}
+ic-cdk-macros = { git = "https://github.com/dfinity/cdk-rs", branch = "main"}
+ic-certified-map = { git = "https://github.com/dfinity/cdk-rs", branch = "main"}
```
Comment (p-shahi) [inline wallet/Cargo.toml:20]: can we add a `rev =` here to pin it to specific revision

File: wallet/src/lib.rs [hunk 5]
```diff
@@ -103,6 +108,109 @@ fn set_name(name: String) {
...
+}
+
+fn init_assets() {
+    let assets = storage::get_mut::<Assets>();
+    for_each_asset(|name, headers, contents, hash| {
+        if name == "/index.html" {
+            assets.hashes.insert("/", *hash);
+            assets
+                .contents
+                .insert("/", (headers.clone().clone().clone(), contents));
```
Comment (roman-kashitsyn) [inline wallet/src/lib.rs:205]: > headers.clone().clone().clone()

I don't quite get why we need to clone 3 times

**Review by roman-kashitsyn — APPROVED** [review]

File: wallet/src/lib.rs (hunk 5 above)
Comment (jplevyak) [reply wallet/src/lib.rs]: vim typeo, fixed

File: wallet/Cargo.toml (hunk 4 above)
Comment (jplevyak) [reply wallet/Cargo.toml]: done

File: dfx.json (hunk 3 above)
Comment (jplevyak) [reply dfx.json]: done

**Comment by jplevyak** [comment]
I tested this on mainnet and it works *except* that when I authorize I need to specify the wallet:

dfx canister --wallet XXXXXXX --network  ic call 'XXXX' authorize '(principal ' ....

of course because I used a wallet to create it as I had to and after authorizing successfully when I try to reload the /authorize URL I get:

Body does not pass verification

I am not sure what is happening.  I think I am getting a 404 but that body doesn't authorize.  @hansl @roman-kashitsyn do I need to do anything special to handle the /authorize redirect ?

**Review by hansl — APPROVED** [review]
LGTM. Thanks!

**Comment by jplevyak** [comment]
bors merge

**Pull request merged**